webapp/src/js/services/rules-engine.js

```diff
@@ -1,256 +1,114 @@
-            fact = new Contact({ contact, reports: [] });
-            facts.push(fact);
-          }
-          fact.reports.push(report);
-        });
-        return facts;
-      };
-
-      var notifyError = function(_err) {
-        err = _err;
-        _.values(callbacks).forEach(function(cbs) {
-          _.values(cbs).forEach(function(callback) {
-            callback(err);
-          });
-        });
-      };
-
-      var notifyCallbacks = function(fact, type) {
-        if (!emissions[type]) {
-          emissions[type] = {};
-        }
-        emissions[type][fact._id] = fact;
-        _.values(callbacks[type]).forEach(function(callback) {
-          callback(null, [ fact ]);
-        });
-      };
-
-      var assertFacts = function() {
-        KNOWN_TYPES.forEach(function(type) {
-          session.on(type, function(fact) {
-            notifyCallbacks(fact, type);
-          });
-        });
-        facts.forEach(function(fact) {
-          session.assert(fact);
-        });
-        session.matchUntilHalt().then(
-          // halt
-          function() {
-            notifyError(new Error('Unexpected halt in fact assertion.'));
-          },
-          // error
-          notifyError
-        );
-      };
+      return Promise.all([ Settings(), UserContact() ])
+        .then(([settingsDoc, userContactDoc]) => {
+          return RulesEngineCore.initialize(DB(), settingsDoc, userContactDoc)
+            .then(() => {
+              const isEnabled = RulesEngineCore.isEnabled();
+              if (isEnabled) {
+                monitorChanges(settingsDoc, userContactDoc);
+              }
 
-      var findFact = function(id) {
-        return _.find(facts, function(fact) {
-          return contactHasId(fact.contact, id) ||
-                 _.findWhere(fact.reports, { _id: id });
+              return isEnabled;
+            });
         });
-      };
-
-      var updateReport = function(doc) {
-        for (var j = 0; j < facts.length; j++) {
-          var fact = facts[j];
-          for (var i = 0; i < fact.reports.length; i++) {
-            if (fact.reports[i]._id === doc._id) {
-              fact.reports[i] = doc;
-              return fact;
-            }
-          }
-        }
-      };
+    });
+  let initialized = initialize();
+
+  const monitorChanges = function (settingsDoc, userContactDoc) {
+    const isReport = doc => doc.type === 'data_record' && !!doc.form;
+    Changes({
+      key: 'mark-contacts-dirty',
+      filter: change => !!change.doc && (ContactTypes.includes(change.doc) || isReport(change.doc)),
+      callback: change => {
+        const subjectId = isReport(change.doc) ? registrationUtils.getPatientId(change.doc) : change.id;
+        RulesEngineCore.updateTasksFor(DB(), subjectId);
+      },
+    });
+
+    const userLineage = [];
+    for (let current = userContactDoc, i = 0; !!current && i < 50; current = current.parent, i++) {
```

Comment:
You don't need `i` because you already know `userLineage.length`. Consider:

```js
for (let current = userContactDoc; !!current && userLineage.length < 50; current = current.parent) {
```

Also, "50" is a magic number so should be a constant not just inlined into a for loop.

Reply:
Nice